src/main/python/deep_qa/training/trainer.py

```diff
@@ -12,10 +12,65 @@
 from ..common.params import get_choice
 from ..data.dataset import Dataset
 from ..data.instances.instance import Instance
-from ..layers.wrappers import OutputMask
 from .models import DeepQaModel
 from .optimizers import optimizer_from_params
 
+# custom layers for use in _get_custom_objects
+from ..layers.adaptive_recurrence import AdaptiveRecurrence
```

Comment:
I would say that it's each class's responsibility to ensure that the layers it uses get into `get_custom_objects`.  It's not reasonable to require that `Trainer` knows about all implemented layers - that breaks encapsulation, and probably leads to circular imports.